zerver/middleware.py

```diff
@@ -421,6 +423,7 @@ def process_response(
             # Note that request.user is a Union[RemoteZulipServer, UserProfile, AnonymousUser],
             # if it is present.
             if hasattr(request, "user") and hasattr(request.user, "format_requestor_for_logs"):
+                assert not isinstance(request.user, AnonymousUser)
```

Comment:
This could just be the `if` condition: `if hasattr(request, "user") and not isinstance(request.user, AnonymousUser)`.

Reply:
Just encountered a django-stubs plugin bug with this one. It only substitutes `AbstractBaseUser -> UserProfile` when there is `AbstractBaseUser | AnonymousUser`, and infers `AbstractBaseUser` when we narrow the type. Should probably also fix this upstream.

Reply:
I rebased this particular change out of this PR as it requires more investigation.